contracts/ROSCA.sol

```diff
@@ -0,0 +1,222 @@
+    membersAddresses.push(msg.sender);
+  }
+
+  /** startRound() check if the round has ended, if so, get the winner of the currentRound's pot
+    * and add the amount to the winner's pending withdrawal
+    * if there were no bid during the round, address at selected at semi-random
+    *
+    * if currentRound = 0(ROSCA hasnt started), no winner is declared
+    */
+
+  function startRound()
+  {
+
+    if (now < startTime + (uint(currentRound + (currentEpoch * membersAddresses.length))  * (uint(roundPeriodInDays) * 1 days)) || membersAddresses.length < minParticipants)
+      throw;
+
+    if (currentRound != 0) {
+      if (winnerAddress == 0) //only true when there is no bidder in this round
+      {
+        // there is no bid in this round so find an unpaid address for this epoch
+        for (uint i = 0 ; i < membersAddresses.length; i++)
+        {
+          uint semi_random = now % membersAddresses.length;
+          if(members[membersAddresses[(currentRound-1 + i) % membersAddresses.length]].paid < currentEpoch)
+          winnerAddress = membersAddresses[currentRound-1 + i];
+          break;
+        }
+      }
+      if(members[winnerAddress].paid == currentEpoch) throw;
+      members[winnerAddress].pendingWithdrawl += lowestBid;
+      members[winnerAddress].paid++;
+      LogRoundFundsReleased(winnerAddress, lowestBid);
+    }
+    if (currentRound < membersAddresses.length)  // reset variables related to bidding
+    {
+      lowestBid = contributionSize * membersAddresses.length;
+      winnerAddress = 0;
+
+      currentRound++;
+    }else if (currentEpoch < numEpochs) // end of Epoch reset rounds
+    {
+      currentEpoch++;
+      currentRound = 1;
+      lowestBid = contributionSize;
+    }
+  }
+
+  /**
+   * Anyone not already a member of ROSCA can request to join and they'll be put into
+   * pendingJoinRequest untill foreman accept request or ROSCA has started
+   *
+   */
+  function joinRequest() onlyBeforeStart
+  {
+    // only put the request in the pending list if they are not in the ROSCA already
+    if (members[msg.sender].alive) throw;
+    pendingJoinRequest[msg.sender] = true;
+    LogParticipantApplied(msg.sender);
+  }
+
+  /**
+   *
+   * once a requestor had been registered as member, the address will be taken out of pendingJoinRequest
+   **/
+  function acceptJoinRequest(address requestor)
+    onlyForeman
+    beforeStart
+  {
+    if (!pendingJoinRequest[requestor]) throw;
+    members[requestor] = USER({paid: 0 , contributed: 0, alive: true, pendingWithdrawl: 0});
+    membersAddresses.push(requestor);
+    numRoundsInEpoch++;
+    LogParticipantApproved(requestor);
+    delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
+  }
+
+  /**
+   * Processes a periodic contribution from msg.sender ().
+   * Any excess funds will be withdrawable through withdraw().
+   */
+  function contribute() payable
+  {
+    if (members[msg.sender].alive)
```

Comment:
we need to throw if (!members[msg.sender].alive), otherwise there'll be funds stuck in the contract